Renamed local_qvm, added options to start quilc and qvm

sauercrowd opened the pull request:
Description
-----------

I renamed `local_qvm` to `local_runtime`, hinting that it does not just start a qvm but also takes care of quilc.

Additionally I added optional parameters that can be provided to that function, giving the user more control over the running processes. 

use_protoQuil defaults to `False`, is that a sensible default?

Currently also exposes options to define custom ports for quilc and qvm, but tbh I'm not sure if that makes sense since (at least right now) I don't see that any other than default ports would work. What do you think?

I also enabled documentation to be generated by this method in sphinx.

Let me know your thoughts on all these and if there are any other options that should get exposed!

Resolves #972 

Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] All new and existing tests pass locally and on Semaphore.
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
- [x] Functions and classes have useful sphinx-style docstrings.
- [x] (New Feature) The docs have been updated accordingly.
- [x] (Bugfix) The associated issue is referenced above using
      [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords).
- [x] The [changelog](https://github.com/rigetti/pyquil/blob/master/CHANGELOG.md) is updated,
      including author and PR number (@username, gh-xxx).

sauercrowd commented:
cc @notmgsk

ecpeterson commented:
I like that this new name better reflects what's happening. I don't like that it's backwards-incompatible.

sauercrowd commented:
Hmmm how about keeping the old `local_qvm` function which just calls on `local_runtime` and adding a deprecation note, eventually removing it from the api?

karalekas commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@ Changelog
     retried only a single time (@notmgsk, gh-951).
 -   The ProtoQuil restrictions built in to PyQVM have been removed
     (@ecpeterson, gh-874).
+-		`local_qvm` has been renamed to `lcocal_runtime` (@sauercrowd, gh-976)
```
```suggestion
-   The `local_qvm` context manager has been renamed to `local_runtime`
    (@sauercrowd, gh-976).
```

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -654,7 +654,13 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
 @contextmanager
-def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+def local_runtime(
+        host: str = '127.0.0.1',
```
these parameters feel like they should be [keyword-only](https://www.python.org/dev/peps/pep-3102/)

notmgsk replied:
I'm unsure how best to do this. Do we list every command line option? just a subset? how do we decide on the subset?

sauercrowd replied:
Parameters are now keyword-only 👍 
I can add all options which would be useful changing (e.g. not expose things like `--verbose` or `log-level`, but all other runtime options) if you would want that, but that may add additional work if options change (especially making sure it's backwards compatible could be difficult)

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -654,7 +654,13 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
 
 @contextmanager
-def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+def local_runtime(
```
I think `local_forest_runtime` is a bit more descriptive

notmgsk replied:
Agreed.

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -654,7 +654,13 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
 
 
 @contextmanager
-def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
```
like @ecpeterson said, let's deprecate this instead of deleting right away

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,22 +675,43 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    :param host: host on which qvm and quilc should listen on
+    :param qvm_port: port which should be used by qvm
+    :param quilc_port: port which should be used by quilc
+    :param provide_http_server: additionally start a http server to the rpcq server (quilc). If true ignores quilc_port and will use 6000 for the http server and 5555 for the rpcq server
```
line too long, can insert a newline somewhere

karalekas commented on pyquil/api/_quantum_computer.py:
```diff
@@ -654,7 +654,13 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+        host: str = '127.0.0.1',
+        qvm_port: int = 5000,
+        quilc_port: int = 5555,
+        provide_http_server: bool = False,
```
@ecpeterson what's the status of this in quilc? are we deprecating the HTTP server?

notmgsk replied:
It is RIP.

sauercrowd commented on pyquil/tests/test_quantum_computer.py:
```diff
@@ -465,7 +465,7 @@ def test_run_and_measure(local_qvm_quilc):
     # note to devs: this is included as an example in the run_and_measure docstrings
     # so if you change it here ... change it there!
-    with local_qvm():  # Redundant with test fixture.
+    with local_runtime():  # Redundant with test fixture.
```
Shoot, thanks! Is it expected that the tests pass even though that function doesn't exist anymore?

sauercrowd commented on docs/source/start.rst:
```diff
@@ -257,20 +257,20 @@ gates from the ``pyquil.gates`` module, which allows us to add operations to our
-    your environment. To make sure both are available you import `from pyquil.api import local_qvm` and then run
-    `local_qvm()`. This will start a qvm and quilc instances using subprocesses if they have not already been started.
+    your environment. To make sure both are available you import `from pyquil.api import local_forest_runtime` and then run
+    `local_forest_runtime()`. This will start qvm and quilc instances using subprocesses if they have not already been started.
```
Maybe try to connect to the specified ports first, and if something is running spit out a warning that it's anticipated that these are qvm/quilc processes. (And if that's supported, request something like the version of quilc/qvm to make sure the processes listening understand the protocol and it's actually qvm/quilc?) Otherwise go ahead and start new instances?

sauercrowd replied:
okay cool, I'll take a look into that

I prototyped sth very quickly, just leave it here incase it's needed at some point in the future

https://gist.github.com/sauercrowd/1a1d7ece4b6dc129700d4126b6aa9e2d

karalekas commented on docs/source/start.rst:
```diff
@@ -257,20 +257,20 @@ gates from the ``pyquil.gates`` module, which allows us to add operations to our
-    your environment. To make sure both are available you import `from pyquil.api import local_qvm` and then run
-    `local_qvm()`. This will start a qvm and quilc instances using subprocesses if they have not already been started.
+    your environment. To make sure both are available you execute ``from pyquil.api import local_forest_runtime`` and then use
+    :py:func:`~pyquil.api.local_forest_runtime()`. This will start qvm and quilc instances using subprocesses if they have not already been started.
```
lets just change the language here and raise a warning (as suggested by @notmgsk and @appleby). otherwise I think this PR is nearly ready to go

sauercrowd replied:
Re the warning: That would mean the processes are started, then wait for maybe two or three seconds, and then check if they exited and raise a warning if necessary? Or should that be a background task and the context manager immediately yields back to the user, while there may appear a warning later?

sauercrowd replied:
@notmgsk @karalekas Let me know if you have suggestions regarding changing the langauge.

Would you like to revert `execute` back to `import` and `use` to `run`?

notmgsk replied:
Looks good to me.

sauercrowd replied:
Ok cool, resolving this 👍

sauercrowd commented:
Just wanted to quickly confirm:
When qvm and quilc are started I can wait for maybe 3 seconds each and then check if the processes exited.
That would add quite a bit of time to the start (I'd say 6 seconds is quite long)

Alternatively I can check if the ports are already in use first, and only start something if they are not (and maybe even then poll until things are available to make sure things are properly running)
If the ports are already used I could issue a warning and leave it to the user.

If I'm not completely wrong the behaviour would be similiar (Since qvm and quilc fail if the ports are unavailable, no matter if these are quilc/qvm or sth else), but would just make the waiting time a bit more dynamic.

If there's a different solution let me know, but since qvm & quilc are both blocking when they run I'm not sure if there's a nice way to ensure they're running properly. (Maybe filtering stdout for keywords? Would bind it to qvm/quilc development though...)

Happy to implement a prototype of the mentioned "check ports" solution, and you can decide if you like it or not!

appleby commented:
> Just wanted to quickly confirm:
> When qvm and quilc are started I can wait for maybe 3 seconds each and then check if the processes exited.
> That would add quite a bit of time to the start (I'd say 6 seconds is quite long)

Well, I suppose the waiting times could coincide, but any guaranteed N-second wait is likely to be painful.

On the other hand, there is something to be said for ensuring that quilc/qvm (or something sufficiently good at impersonating them!) really are running and responding on the requested ports before returning...

My only hesitation is that polling will add some complexity, and the current approach of just blindly carrying on seems to work quite well in practice! So maybe it's not worth doing.

But I wouldn't be opposed if you want to give the polling a shot.  Just be prepared to revert in case other reviewers feel differently :)

notmgsk commented:
Apologies for the late response. I don't think we need to introduce any delay. @sauercrowd I like the method you outlined [here](https://gist.github.com/sauercrowd/1a1d7ece4b6dc129700d4126b6aa9e2d). The only change I would make is to skip the "check if the thing that is listening to port 5000 is a qvm service" part. If there is something listening on that port, then just assume it's a qvm server: if it isn't, an appropriate error will bubble up. Likewise quilc.

karalekas commented:
As a heads up, you're going to want to rebase off of master and move your changelog entry to the new v2.12 section!

sauercrowd commented:
I added a small function `_port_used` to check if a process is already listening on a port. Other than that I just added a small if statement for both the quilc and the qvm process, and issue a warning on stderr if the port is already in use.

Let me know what you think!

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+               "port {} is already in use.").format(qvm_port), file=sys.stderr)
+    else:
+        qvm_cmd = ['qvm', '-S', '--host', host, '-p', str(qvm_port)]
+        qvm = subprocess.Popen(qvm_cmd,
```
Both qvm and quilc have a non-zero startup time (on the order of 1s). If `subprocess.Popen` does not block until the process completes, then we could have the situation where the user makes a request to the servers before they are truly available.

I suggest we call `wait()` on both `qvm` and `quilc`.

```
In [10]: %time p = subprocess.Popen("sleep 20", shell=True)
CPU times: user 435 µs, sys: 3.11 ms, total: 3.55 ms
Wall time: 3.06 ms

In [11]: %time p.wait()
CPU times: user 1.91 ms, sys: 1.65 ms, total: 3.56 ms
Wall time: 19.3 s
Out[11]: 0
```

ecpeterson replied:
For what it's worth: unless pyQuil is explicitly planning to get upset about a missing / timing-out quilc (as in `QuantumComputer` initialization), it is OK to send an RPCQ client message before there's a waiting RPCQ server. The ZMQ client will hold on to the message 'til there's a server ready to receive it, and the user won't notice anything but blocking.

notmgsk replied:
That's good to know! I'm not sure how qvm will react however. If ZMQ will block anyway, might as well do that upfront here for completeness' sake.

appleby replied:
I might be misunderstanding, but `Popen.wait` blocks until the process terminates, unless you pass it a `timeout` arg. If we block on a wait call, we can't return to the caller.

You could always pass a timeout as a way to try and give the process some time to startup, but you still don't really know if it's up and ready to receive requests unless you actually send a request to the port, similar to @sauercrowd's previous suggestion.

ecpeterson replied:
Yeah: my comment doesn't apply to the QVM. Python `requests` will get upset that there's no listening HTTP server.

notmgsk replied:
@appleby I was thinking that qvm was going to daemonize itself after it finished setting up. I dunno why I was thinking that. The issue remains.

sauercrowd replied:
Can we wait for qvm/quilc in subprocess? Won't they keep on blocking?

I implemented sth a couple of days ago which polls until until qvm and quilc are available, happy to implement it that way here if you think that makes sense

https://github.com/sauercrowd/rqubit/blob/0b81231d75c86cb5fb0d936b37e9b6cc21ecce4d/rqubit/quomputer.py#L50

notmgsk replied:
Darn. This is messier than I was expecting. 

@sauercrowd looks like we will have to do some polling like in your example code. Unless @appleby @ecpeterson have any other suggestions, then let's add a `poll_timeout=2` and `poll_interval=0.2`, and check whether the servers are available within `poll_timeout`.

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    .. warning::
+        If ``use_protoquil`` is set to ``True`` language features you need
+        may be disabled. Please use it with caution.
+
```
Document the returned tuple. Explain what a value of `None` in the tuple means.

notmgsk commented on CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ Changelog
 
 ### Improvements and Changes
+-   The `local_qvm` context manager has been renamed to `local_forest_runtime`,
+    which now checks if the designated ports are used before starting `QVM`/`QUILC`.
```
lowercase qvm and quilc

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,6 +671,40 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+
+def _port_used(port: int):
+    """Check if a (TCP) port is listening on localhost.
+    Returns ``True`` if a process is listening on the specified port, ``False`` otherwise.
```
Use the `:return ...: ...` docstr format.

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,6 +671,40 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+    """Check if a (TCP) port is listening on localhost.
+    Returns ``True`` if a process is listening on the specified port, ``False`` otherwise.
+
+    :param port: TCP port which should get checked.
```
"TCP port to check."

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,6 +671,40 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.connect(('127.0.0.1', port))
```
We ought to make the host configurable, since it is configurable in `local_forest_runtime`.

sauercrowd replied:
Just to clarify, I may got that wrong.

The `host` parameter in `local_forest_runtime` defines which address to listen on,
so if we try to connect to `127.0.0.1` we should always hit what we want to hit no?

More than happy though to make that configurable and pass `127.0.0.1` from the `local_forest_runtime` as a parameter, I think that'd make sense!

sauercrowd replied:
I added the host parameter for the `_check_port` function, and `local_forest_runtime` passes now `127.0.0.1` to them. If this is not what's needed happy to revert it back/change it.

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,6 +671,40 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+        qvm_port: int = 5000,
+        quilc_port: int = 5555,
+        use_protoquil: bool = False) -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    """A context manager for the Rigetti local QVM and QUIL compiler.
```
"A context manager for local qvm and quilc processes."

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    quilc = None
+
+    if _port_used(qvm_port):
+        print(("Unable to start QVM, since the QVM "
```
"Unable to start qvm server, since the specified port {} is in use."

notmgsk replied:
Make this a warning.

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE)
+    if _port_used(quilc_port):
+        print(("Unable to start QUILC, since the QUILC "
```
Unable to start the quilc server, since the specified port {} is in use.

notmgsk replied:
Make this a warning.

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+        if use_protoquil:
+            quilc_cmd += ['-P']
+
+        quilc = subprocess.Popen(quilc_cmd,
```
See comment RE `wait()`.

sauercrowd commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,45 +671,114 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
-    This context manager will start up external processes for both the
-    compiler and virtual machine, and then terminate them when the context
-    is exited.
+    This context manager will ensure that the designated ports are not used, start up `qvm` and
```
@notmgsk Documented the new behaviour and the tuple values here. I wasn't sure if I should use `:return:` since this is an iterator.

Looking forward to your feedback!

notmgsk replied:
If it's not standard to use `:return:` for an iterator, then what we have now is fine.

sauercrowd replied:
Added a return section, resolving this

appleby commented on pyquil/api/_quantum_computer.py:
```diff
@@ -14,6 +14,8 @@
 ##############################################################################
 import re
+import socket
+import sys
```
I think you can drop this `sys` import now, no?

sauercrowd replied:
Good catch!

appleby commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,45 +671,114 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+    qvm = None
+    quilc = None
+
+    if _port_used('127.0.0.1', qvm_port):
```
I think you should pass `host` here, rather than always passing `127.0.0.1`. Likewise for quilc below.

sauercrowd replied:
I may be wrong, but the host we're accepting as a parameter is a server side argument, no?
E.g. someone may pass `0.0.0.0` to accept connections from all interfaces, in contrast to the host we pass to `_port_used`, which is the client-side parameter to specify which address to connect to.

Please correct me if I'm wrong, may be heading in the complete wrong direction here.

appleby replied:
You are correct, `host` here is the address the server will bind. The case of what to do with "0.0.0.0" is a good point. But if the caller specifies anything other than 0.0.0.0 or 127.0.0.1, then you don't necessarily want to query 127.0.0.1.

Maybe if they pass 0.0.0.0, check localhost; otherwise `host`?

appleby replied:
Also, for what it's worth, it common for implementations to allow you to `connect(2)` to 0.0.0.0, even though I think it's technically only valid as a source address?

Anyway, I just tested an Python on my macbook doesn't complain, and essentially treats it as if I had specified localhost, so maybe we don't have to do anything special... not sure how portable / reliable that is though.

sauercrowd replied:
> Also, for what it's worth, it common for implementations to allow you to connect(2) to 0.0.0.0, even though I think it's technically only valid as a source address?

Interesting, didn't know that!

Just realised that it's possible to listen to the same port but bind to different addresses, so there may be a scenario where the user has an application e.g. listening to `127.0.0.1` on port 5000, but when binding to a different address, let's say `10.11.12.102`, we can still listen to port 5000 for qvm. (Summing up things to make sure I got it right)

Long story short, it completely makes sense to pass the host to `_check_port`. Happy to add an additional `if` to replace `0.0.0.0` with `127.0.0.` to be on the safe side?

Let me know!

appleby replied:
I think passing `host` and mapping 0.0.0.0 --> 127.0.0.1 is a reasonable compromise.

Even that is not technically correct, since it's not required that localhost is available or even that a loopback interface is configured, but it seems reasonable to assume 0.0.0.0 will bind to 127.0.0.1 in most environments were pyquil is running.

sauercrowd replied:
Added!

sauercrowd commented:
I think what's outstanding now is the updated docstring and how to ensure that everything is up and running so `requests` won't complain

notmgsk commented on pyquil/api/_quantum_computer.py:
```diff
@@ -669,45 +670,116 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+    qvm = None
+    quilc = None
+
+    # if host is 0.0.0.0 replace it with 127.0.01
```
Rather than documenting *what* is happening (the code does that already), document *why*.

notmgsk commented:
@sauercrowd After some out-of-band discussion with the gang, we've agreed that the polling stuff introduces additional complexity that maybe isn't necessary for this PR. Indeed, polling was not needed in the original implementation of this feature, and so it likely won't be an issue if we ignore it altogether in this PR.

The rest of the PR is great though. So let's nix the polling stuff, and address the remaining issues (if there are any).

sauercrowd commented:
Sounds good!

The only thing from my end would be the `:return:` you mentioned, it seems to be fine to document iterators/generators in it as well.

So if you're okay with it I'd move a bit of the documentation of `local_forest_runtime` to a `:return:` section.

notmgsk commented:
> Sounds good!
> 
> The only thing from my end would be the `:return:` you mentioned, it seems to be fine to document iterators/generators in it as well.
> 
> So if you're okay with it I'd move a bit of the documentation of `local_forest_runtime` to a `:return:` section.

Go for it

sauercrowd commented:
I think things are addressed now, should I re-request review from all people involved?

sauercrowd commented on docs/source/start.rst:
```diff
@@ -257,20 +257,20 @@ gates from the ``pyquil.gates`` module, which allows us to add operations to our
-        from pyquil.api import local_qvm
+        from pyquil.api import local_forest_runtime
 
         qvm = get_qc('9q-square-qvm')
```
moved it into the with block 👍

sauercrowd commented:
Thanks you all for the feedback and patience!
